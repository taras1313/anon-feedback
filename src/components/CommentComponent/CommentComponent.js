import React, { Component } from 'react';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import withStyles from '@material-ui/core/styles/withStyles';
import Chip from '@material-ui/core/Chip';
import cx from 'classnames';

import { ThreadAuthorComponent } from '../ThreadAuthorComponent';
import { ThreadAuditComponent } from '../ThreadAuditComponent';

import styles from './CommentComponent.module.scss';

const classes = () => ({
  chip: {
    marginRight: '6px',
    height: '26px'
  }
});

class Comment extends Component {
  state = {
    isEditing: false,
    text: '',
  };

  toggleIsEditing = () => {
    const { isEditing, text: newCommentText } = this.state;
    const { comment: { text, _id: commentId }, onUpdateComment } = this.props;

    if (!isEditing) { // open editing
      this.setState({ isEditing: !isEditing, text })
    } else { // saving
      onUpdateComment({ text: newCommentText, commentId }).then(() => {
        this.setState({ isEditing: !isEditing, text: '' })
      });
    }
  };

  editComment = ({ target: { value } }) => this.setState({ text: value });

  resetIsEditing = () => this.setState({ isEditing: false });

  renderButtons = () => {
    const { editable } = this.props;
    if (!editable) return null;

    const { isEditing } = this.state;

    return (
      <div className={isEditing ? styles.btnHolder : ''}>
        {isEditing && (
          <Button onClick={this.resetIsEditing} size='small'>
            Cancel
          </Button>
        )}

        <Button onClick={this.toggleIsEditing} color="secondary" size="small">
          {isEditing ? 'Save' : 'Edit'}
        </Button>
      </div>
    );
  };

  constructAdditionalText = () => {
    const { comment: { wasEdited }, isAuthorsComment } = this.props;
    let text = '';
    if (isAuthorsComment) {
      text = '@thread author ';
    }

    if (wasEdited) {
      text += '(edited)'
    }

    return text.trim();
  };

  repliedToHandler = () => {
    const { repliedToHandler, comment: { author: { username } } } = this.props;

    repliedToHandler(username);
  };

  render() {
    const { comment: { text, author: { username }, repliedToUser }, userId, classes } = this.props;
    const { isEditing, text: inputText } = this.state;

    return (
      <div className={styles.comment}>
        <div className={styles.flexWrapper}>
          <ThreadAuthorComponent
            username={username}
            label={this.constructAdditionalText()}
          />
          {this.renderButtons()}
        </div>

        <div className={cx(styles.text, styles.leftShift)}>
          {!!repliedToUser && !isEditing && (
            <Chip
              variant="outlined"
              className={classes.chip}
              label={`@${repliedToUser.username}`}
              color={userId === repliedToUser.userId ? 'secondary' : 'primary'}
            />
          )}
          {!isEditing && text}

          {isEditing && (
            <TextField
              autoComplete='off'
              id="standard-multiline-flexible"
              label="Edit your comment"
              multiline
              value={inputText}
              fullWidth
              onChange={this.editComment}
              // margin="normal"
              variant="outlined"
              // variant="outlined" // ?
            />
          )}
        </div>

        <div className={cx(styles.flexWrapper, styles.leftShift)}>
          <ThreadAuditComponent />
          <Button color='primary' onClick={this.repliedToHandler} variant='outlined' size='small'>
            Reply
          </Button>
        </div>
        <div className={styles.divider} />
      </div>
    );
  }
}

export const CommentComponent = withStyles(classes)(Comment);