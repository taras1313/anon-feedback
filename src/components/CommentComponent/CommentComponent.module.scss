.divider {
  width: 100%;
  background: #dcdcdc;
  height: 1px;
  margin: 20px 0;
}

.flexWrapper {
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.comment {
  padding-top: 16px;
}

.text {
  margin: 10px 0;
  line-height: 22px;
  font-size: 15px;
  white-space: pre-line;
}

.leftShift {
  padding-left: 26px;
}

.btnHolder {
  width: 140px;
  display: flex;
  justify-content: space-between;
}