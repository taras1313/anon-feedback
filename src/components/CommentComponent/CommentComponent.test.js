import React from "react";
import { shallow } from "enzyme";
import { Comment } from "./CommentComponent";
import Button from "@material-ui/core/Button";
import TextField from "@material-ui/core/TextField";

const props = {
  threadId: "1",
  userId: "2",
  likeComment: () => {},
  dislikeComment: () => {},
  comment: {
    likesList: ["2"],
    dislikesList: ["1"],
    _id: "3",
    text: "mockTest",
    likesCount: "10",
    dislikesCount: "20",
    createdDate: "01/20/2010",
    author: { username: "smallchanov" },
    repliedToUser: "tarasnormas",
  },
  editable: true,
  repliedToHandler: jest.fn(),
  classes: {},
};

describe("comment component", () => {
  test("snapshot", () => {
    const wrapper = shallow(<Comment {...props} editable={false} comment={{...props.comment, wasEdited: true}} />);

    expect(wrapper).toMatchSnapshot();
  });

  test("dislike status", () => {
    shallow(<Comment {...props} dislikesList={["2"]} />);
  });

  test("toggle editing", () => {
    const wrapper = shallow(<Comment {...props} />);

    wrapper.find(Button).at(0).props().onClick();
    expect(wrapper.instance().state.isEditing).toBe(true);

    wrapper
      .find(TextField)
      .props()
      .onChange({ target: { value: "testValue" } });
    expect(wrapper.instance().state.text).toBe("testValue");

    wrapper.find(Button).at(0).props().onClick();
    expect(wrapper.instance().state.isEditing).toBe(false);
  });

  test('reply',  () => {
    const wrapper = shallow(<Comment {...props} isAuthorsComment />);

    wrapper.find(Button).at(1).props().onClick();

    expect(props.repliedToHandler).toHaveBeenCalledWith('smallchanov');
  });
});
