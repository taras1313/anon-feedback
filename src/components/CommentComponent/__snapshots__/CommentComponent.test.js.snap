// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comment component snapshot 1`] = `
<div
  className="comment"
>
  <div
    className="flexWrapper"
  >
    <ThreadAuthorComponent
      label="(edited)"
      username="smallchanov"
    />
  </div>
  <div
    className="text leftShift"
  >
    <WithStyles(ForwardRef(Chip))
      color="primary"
      label="@undefined"
      variant="outlined"
    />
    mockTest
  </div>
  <div
    className="flexWrapper leftShift"
  >
    <ThreadAuditComponent
      createdDate="01/20/2010"
      dislike={[Function]}
      dislikesCount="20"
      isClickable={true}
      like={[Function]}
      likeStatus="liked"
      likesCount="10"
    />
    <WithStyles(ForwardRef(Button))
      color="primary"
      onClick={[Function]}
      size="small"
      variant="outlined"
    >
      Reply
    </WithStyles(ForwardRef(Button))>
  </div>
  <div
    className="divider"
  />
</div>
`;
